pkg/util/objectwatcher/objectwatcher.go

```diff
@@ -97,6 +101,30 @@ func (o *objectWatcherImpl) Create(clusterName string, desireObj *unstructured.U
 	return nil
 }
 
+func (o *objectWatcherImpl) retainClusterFields(desired, observed *unstructured.Unstructured) error {
+	// Pass the same ResourceVersion as in the cluster object for update operation, otherwise operation will fail.
+	desired.SetResourceVersion(observed.GetResourceVersion())
+
+	// Retain finalizers since they will typically be set by
+	// controllers in a member cluster.  It is still possible to set the fields
+	// via overrides.
+	desired.SetFinalizers(observed.GetFinalizers())
+	// Merge annotations since they will typically be set by controllers in a member cluster
+	// and be set by user in karmada-controller-plane.
+	util.MergeAnnotations(desired, observed)
+
+	if o.resourceExplorer.HookEnabled(desired, configv1alpha1.ExploreRetaining) {
+		obj, err := o.resourceExplorer.Retain(desired, observed)
+		if err != nil {
+			return err
+		}
+		*desired = *obj
```

Comment:
That will incur an unnecessary memory copy.
Consider:
```diff
diff --git a/pkg/util/objectwatcher/objectwatcher.go b/pkg/util/objectwatcher/objectwatcher.go
index 34f93696..52e7055e 100644
--- a/pkg/util/objectwatcher/objectwatcher.go
+++ b/pkg/util/objectwatcher/objectwatcher.go
@@ -101,7 +101,7 @@ func (o *objectWatcherImpl) Create(clusterName string, desireObj *unstructured.U
        return nil
 }

-func (o *objectWatcherImpl) retainClusterFields(desired, observed *unstructured.Unstructured) error {
+func (o *objectWatcherImpl) retainClusterFields(desired, observed *unstructured.Unstructured) (*unstructured.Unstructured, error) {
        // Pass the same ResourceVersion as in the cluster object for update operation, otherwise operation will fail.
        desired.SetResourceVersion(observed.GetResourceVersion())

@@ -114,15 +114,10 @@ func (o *objectWatcherImpl) retainClusterFields(desired, observed *unstructured.
        util.MergeAnnotations(desired, observed)

        if o.resourceExplorer.HookEnabled(desired, configv1alpha1.ExploreRetaining) {
-               obj, err := o.resourceExplorer.Retain(desired, observed)
-               if err != nil {
-                       return err
-               }
-               *desired = *obj
-               return nil
+               return o.resourceExplorer.Retain(desired, observed)
        }

-       return nil
+       return desired, nil
 }

 func (o *objectWatcherImpl) Update(clusterName string, desireObj, clusterObj *unstructured.Unstructured) error {
@@ -138,7 +133,7 @@ func (o *objectWatcherImpl) Update(clusterName string, desireObj, clusterObj *un
                return err
        }

-       err = o.retainClusterFields(desireObj, clusterObj)
+       desireObj, err = o.retainClusterFields(desireObj, clusterObj)
        if err != nil {
                klog.Errorf("Failed to retain fields for resource(kind=%s, %s/%s) in cluster %s: %v", desireObj.GetKind(), desireObj.GetNamespace(), desireObj.GetName(), clusterName, err)
                return err
```